### PR TITLE
Layer View 3D, with detection of OpenGL context versions.

### PR DESCRIPTION
**Pull request by jackha**
Summary:
- Added possibility for arbitrary attributes to MeshData
- Merged interfaces of UM.View.GL and UM.Qt.GL by moving all code from UM.Qt.GL to UM.View.GL. This is the majority of code changes.

OpenGL 4.1 core and 2.0
- Added OpenGLContext and OpenGLContextProxy
- Added 41core shaders.
- Use 41core shader if supported, else use compatibility shader.
- VertexArrayObjects are used if supported
- Added context compatibility check, for 4.1 core and also for 2.0
- Try to set 4.1 core profile, if failed, take 2.0, if failed error

### CODE REVIEW

File: UM/Settings/ContainerStack.py [hunk 1]
```diff
@@ -158,7 +158,8 @@ def setDirty(self, dirty):
     def getProperty(self, key, property_name):
         value = self.getRawProperty(key, property_name)
         if isinstance(value, SettingFunction.SettingFunction):
-            return value(self)
```
Comment (nallath) [inline UM/Settings/ContainerStack.py]: Uh. What is the point of this change?

Comment (jackha) [inline UM/Settings/ContainerStack.py]: Ahhh used it with debugging, I will revert it.

File: UM/View/GL/Texture.py [hunk 2]
```diff
@@ -1,32 +1,52 @@
...
+            elif self._image is None: # No filename or image set.
+                self._image = QImage(1, 1, QImage.Format_ARGB32)
+                self._image.fill(0)
+            self._qt_texture.setData(self._image)
+            self._qt_texture.setMinMagFilters(QOpenGLTexture.Linear, QOpenGLTexture.Linear)
 
+        self._qt_texture.bind(unit)
     ##  Release the texture from a certain texture unit.
     #
     #   \param texture_unit The texture unit to release from.
```
Comment (nallath) [inline UM/View/GL/Texture.py]: Documentation doesn't match param name (it's now called unit). I do kinda prefer the texture_unit name, as unit is a bit confusing.

File: UM/Qt/Bindings/OpenGLContextProxy.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2015 Ultimaker B.V.
```
Comment (jackha) [inline UM/Qt/Bindings/OpenGLContextProxy.py]: Apparently I've been living in the past

File: UM/View/GL/OpenGL.py [hunk 4]
```diff
@@ -52,25 +102,32 @@ def getGPUType(self):
...
 
     ##  Create a ShaderProgram Object.
     #
     #   This should return an implementation-specifc ShaderProgram subclass.
     def createShaderProgram(self, file_name):
-        raise NotImplementedError("Should be implemented by subclasses")
+        shader = ShaderProgram.ShaderProgram()
+        # The version_string must match the keys in shader files.
+        if OpenGLContext.isLegacyOpenGL():
+            version_string = ""
```
Comment (jackha) [inline UM/View/GL/OpenGL.py]: Ahh you mean this version_string... yes agree, it's more explicit this way.

File: UM/View/GL/OpenGL.py [hunk 5]
```diff
@@ -19,31 +32,68 @@ class Vendor:
         Intel = 3
         Other = 4
 
+    def __init__(self):
+        profile = QOpenGLVersionProfile()
```
Comment (jackha) [inline UM/View/GL/OpenGL.py]: What do you mean? The profile object is created to get versionFunctions. The Qt API is somewhat strange here IMHO.

File: UM/View/GL/OpenGL.py [hunk 6]
```diff
@@ -86,7 +143,66 @@ def createShaderProgram(self, file_name):
...
     def createVertexBuffer(self, mesh, **kwargs):
-        raise NotImplementedError("Should be implemented by subclasses")
+        if not kwargs.get("force_recreate", False) and hasattr(mesh, OpenGL.VertexBufferProperty):
+            return getattr(mesh, OpenGL.VertexBufferProperty)
+
+        buffer = QOpenGLBuffer(QOpenGLBuffer.VertexBuffer)
+        buffer.create()
+        buffer.bind()
+
+        buffer_size = mesh.getVertexCount() * 3 * 4 # Vertex count * number of components * sizeof(float32)
```
Comment (jackha) [inline UM/View/GL/OpenGL.py]: Do you mean sys.getsizeof or ctypes.sizeof ? Didn't manage to get this "4" back in an easy way.

Comment (jackha) [inline UM/View/GL/OpenGL.py]: ctypes.sizeof(ctypes.c_float)

File: UM/View/GL/OpenGLContext.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    @classmethod
+    def detectBestOpenGLVersion(cls):
+        ctx = cls.setContext(4, 1, core = True)
+        if ctx is None:
+            return None, None, None
+        fmt = ctx.format()
+        profile = fmt.profile()
+
+        # First test: we hope for this
+        if fmt.majorVersion() >= 4 and fmt.minorVersion() >= 1 and profile == QSurfaceFormat.CoreProfile:
```
Comment (jackha) [inline UM/View/GL/OpenGLContext.py]: Well that's what the code says ^^. Fixed

File: UM/View/GL/Texture.py (hunk 2 above)
Comment (jackha) [inline UM/View/GL/Texture.py]: Ah I copied over the function, apparently the parameter was called differently. Fixed it.

File: UM/View/GL/OpenGL.py (hunk 5 above)
Comment (nallath) [inline UM/View/GL/OpenGL.py]: Derp. I wasn't paying attention. Disregard the comment.

File: UM/View/GL/OpenGL.py (hunk 4 above)
Comment (jackha) [inline UM/View/GL/OpenGL.py]: Actually, I like it how it is currently. "fragment" and "vertex" are the baseline, and "fragment41core", "vertex41core", "geometry41core" are the special ones. I've added a comment to make it clearer.

**Pull request merged**